### DISCUSSION TITLE
How to dynamically manage a playlist #1043

### DISCUSSION THREAD

**Discussion by carrzkiss** [savonet/liquidsoap#3860] [General]
Back in 2019, [tristan-nel](https://github.com/tristan-nel) asked the question, `How to dynamically manage a playlist`.
On Mar 2, 2020, he provided the code on how to do it, but for those who have never used Python scripting language, I am at a loss on what to do with it to use it.

This was his final post, and it was cool that he returned and posted it.
===Final Post===
So I ended up using `request.dynamic`, and using a Python script to request what the next song in the user-managed playlist will be from Cloud Firestore:

In the liquidsoap server:
```
# Functions
def get_request() =
  uri = list.hd(default="", get_process_lines("python3 request.py"))
  request.create(uri, persistent=true)
end

# Requests
requests = request.dynamic(
  id="requests",
  conservative=false,
  get_request
)
```
In the Python script:

```
# ... get the next song info from Firestore
URL = next_song[const.URL]
new_now_playing = now_playing + 1
channel_ref.set({const.NOW_PLAYING : new_now_playing}, merge=True)
print(URL)
```
===End Final Post===

I am going to guess here on what to do.
I would have to install `python3` (After running the below command, it lets me know I don't have it installed.)
```
liquid@linuxserver: python –version
  Command 'python' not found, did you mean:
  command 'python3' from deb python3
  command 'python' from deb python-is-python3
```
The Python script would need to be saved as `request.py` in the same folder as the `.liq` file.
The code in the liq file is similar to the example code on [Request.dynamic](https://www.liquidsoap.info/doc-dev/request_sources.html#request.dynamic).

Will this make it look at the playlist file for new entries, or will you have to do some injection to get the new songs added to the queue?

**Comment by vitoyucepi** [comment]
`python3` should be installed, check `python3 --version` or `python3 -V`.

```
def get_request()
  command = process.quote.command("python3", args=["request.py"])
  uri = list.hd(process.read.lines(command), default="")
  request.create(uri, persistent=true)
end
```

Also, there is a bug #3134 that could lead to an insatiable hunger mode for `request.dynamic`.

```
def get_request()
  command = process.quote.command("python3", args=["request.py"])
  uri =
    try
      list.hd(process.read.lines(command))
    catch _ do
      null()
    end

  null.map(fun(uri) -> request.create(uri, temporary=true), uri)
end

def replaces request.dynamic(%argsof(request.dynamic), callback)
  prevent = ref(false)
  def debounce()
    if prevent() then
      null()
    else
      result = callback()
      if result == null() then
        thread.run({prevent := false}, delay=retry_delay, fast=true)
        prevent := true
      end
      result
    end
  end

  request.dynamic(debounce, %argsof(request.dynamic))
end


requests = request.dynamic(get_request, id="requests", retry_delay=10.)
```

**Reply by carrzkiss** [reply]
OK, I did not think about checking for that specific version.
Python 3.10.12

OK, I will keep an eye on this thread, and once the bug fix is resolved, I will proceed from there.

**Reply by vitoyucepi** [reply]
Update to `request.dynamic`. I've found out that hunger mode can be activated when the requests are invalid. So here's the solution
```ruby
def replaces request.dynamic(%argsof(request.dynamic), callback)
  prevent = ref(false)

  def disable_prevent()
    prevent := false
  end

  def enable_prevent()
    prevent := true
    delay = max(getter.get(retry_delay) - 0.1, 0.1)
    thread.run(disable_prevent, delay=delay)
  end

  calls = ref(0)
  max_calls = 3

  def reset_calls()
    calls := 0
  end

  def increment_calls()
    calls := calls() + 1
  end

  thread.run(reset_calls, every=retry_delay, fast=true)

  def debounce()
    if prevent() then
      null()
    elsif calls() >= max_calls then
      null()
    else
      result = callback()
      if result == null() then
        enable_prevent()
      end
      increment_calls()
      result
    end
  end

  request.dynamic(debounce, %argsof(request.dynamic))
end
```

P.S. It is much better to return `null()` than `request.create("")`, because the latter is invalid.

**Reply by Moonbase59** [reply]
Now _that’s_ an interesting piece of code @vitoyucepi! "Hunger mode", I’m still grinning…

**Reply by vitoyucepi** [reply]
Hi @Moonbase59,
`request.dynamic` has no safeguards, if misconfigured it can generate an enormous amount of requests.
I think the correct term would be self-employed DoS. If you are using something in the cloud, you will of course have cloud limits for cloud API keys. The average song length is about 180 seconds, I'm able to produce 25 requests per second at 100% CPU. The question is, how fast would the API key be banned?

**Reply by Moonbase59** [reply]
> request.dynamic has no safeguards, if misconfigured it can generate an enormous amount of requests.

Something important to remember, thanks!

**Reply by carrzkiss** [reply]
Where would this be plugged in at?
Would it be before or after
`radio101 = mksafe(crossfade(fade_out=3., fade_in=3., duration=5., radio101))`

**Reply by carrzkiss** [reply]
I re-read what you wrote about the constant request and 100% CPU. Is this code you created safe to use, or should I wait a bit longer?
I am also trying to see how it would be used. I looked at the linked thread and am trying to work it out.

This is my playlist.
`music = playlist(mode='normal',reload=1, reload_mode='rounds', "https://demo.example.com/DJ/88_1.m3u")`
and this is my `station` and its `mksafe`
``` 
radio101 = metadata.map(update=false, hls_meta,music)
radio101 = mksafe(crossfade(fade_out=3., fade_in=3., duration=5., radio101))
```

By looking at the other thread, I see that he has `request.dynamic()` wrapped around his jukebox,

Also, looking at the code.
Is there a way to set a timer on it so that it will only check after, say, 30 seconds or less or longer?
That would definitely slow down the violent request on the CPU.

Take care, Vito

**Reply by vitoyucepi** [reply]
> I re-read what you wrote about the constant request and 100% CPU. Is this code you created safe to use, or should I wait a bit longer?

It's okay, you can use it.

> Is there a way to set a timer on it so that it will only check after, say, 30 seconds or less or longer?

`requests = request.dynamic(get_request, id="requests", retry_delay=30.)`
10 seconds should be fine, I guess, compared to 0.04 seconds, by the way, what are the limits for your cloud requests?

**Reply by carrzkiss** [reply]
Do I have to add the python script as well or is this all done through your script?

Cloud--------
Local. Running my own servers.
The vm is running on a 
Dell r710 
I have another Dell that I am going to dedicate to the Linux vms later on. Right now, I am just getting this one vm running, then I will bring the other Dell online and throw a few liquidsoap servers on it.

**Reply by carrzkiss** [reply]
This is just added to the liq script and not wrapped around anything else.
`requests = request.dynamic(get_request, id="requests", retry_delay=30.)`
Will it go before or after this part?
```
radio101 = metadata.map(update=false, hls_meta,music)
radio101 = mksafe(crossfade(fade_out=3., fade_in=3., duration=5., radio101))
```
I will test it out tomorrow or early morning, whichever comes first.
Have a good one, Vito.

**Reply by vitoyucepi** [reply]
> Will it go before or after this part?

You can use the *requests* snippet before this part.

> `requests = request.dynamic(get_request, id="requests", retry_delay=30.)`

That's not all, you may (or may not) want to use my custom `request.dynamic` snippet.
If you want to merge the requests with the music, you can use `fallback([requests, music])`, `switch`, `rotate` or `random`.

**Reply by carrzkiss** [reply]
OK, I had to look up how to use the `fallback` as I had never used it before.
I am unsure if this will be the correct way to use it, but here we go.

The dynamic is in its own file.
```
%include "plugins/dynamic.liq"
 radio101 = fallback([requests, music])
 radio101 = metadata.map(update=false, hls_meta,music)
 radio101 = mksafe(crossfade(fade_out=3., fade_in=3., duration=5., radio101))
```

I am right; it is not the right way to do it.

```
Error 4: Undefined variable requests
radio101 = fallback([requests, music])
```

**Reply by vitoyucepi** [reply]
> ```
> Error 4: Undefined variable requests
> radio101 = fallback([requests, music])
> ```

>> `requests = request.dynamic(get_request, id="requests", retry_delay=30.)`

**Reply by carrzkiss** [reply]
This is my code now.
```
%include "plugins/dynamic.liq"
requests = request.dynamic(get_request, id="requests", retry_delay=30.)
radio101 = fallback([requests, music])
radio101 = metadata.map(update=false, hls_meta,music)
```
Restarted LS, and it immediately stopped.
Logs show this.
```
liquidsoap@101.service: Failed with result 'exit-code'.
liquidsoap@101.service: Main process exited, code=exited, status=1/FAILURE
Error 4: Undefined variable get_request
requests = request.dynamic(get_request, id="requests", retry_delay=30.)
At /home/liquid/liquidsoap/101.liq, line 49, char 27-38:
```

`get_request` is from this code [9016807](https://github.com/savonet/liquidsoap/discussions/3860#discussioncomment-9016807)
The new code does not have such a variable. [9018614](https://github.com/savonet/liquidsoap/discussions/3860#discussioncomment-9018614)
OR do both codes go together?

Nope----

**Reply by vitoyucepi** [reply]
> ```
> Error 4: Undefined variable get_request
> requests = request.dynamic(get_request, id="requests", retry_delay=30.)
> At /home/liquid/liquidsoap/101.liq, line 49, char 27-38:
> ```

Liquidsoap doesn't see `get_request`.

> ```ruby
> %include "plugins/dynamic.liq"
> requests = request.dynamic(get_request, id="requests", retry_delay=30.)
> radio101 = fallback([requests, music])
> radio101 = metadata.map(update=false, hls_meta, music)
> ```

Neither do I.

**Reply by carrzkiss** [reply]
I just needed to go to bed and wake up still not fully rested for me to read
`Neither do I.`
For me to notice that I needed to take the top of `9016807` and put it on top of `9018614` and it went to working.
Thanks, Vito.
I am testing it right now.
I moved a song, and it did not pick it up. Then, after two songs played, I moved another song, and it picked it up and played it immediately after the song that was playing finished.
So, it seems to be working.
Thank you so much, Vito, and thanks for the simple comment, `Neither do I,` as that is what I needed to see what I was missing.

**Reply by carrzkiss** [reply]
It plays the current list to the end, then picks up and starts playing the newly added songs.
Am I missing something? Isn't this supposed to pick up on newly added songs and play them next?
I changed the check to 10 seconds to see if that was the issue, but it was not.

My Code.

```
music = playlist(mode='normal',reload=1, reload_mode='rounds', "https://demo.example.com/DJ/88_1.m3u")
%include "plugins/dynamic.liq"
requests = request.dynamic(get_request, id="requests", retry_delay=10.)
radio101 = fallback([requests, music])
radio101 = metadata.map(update=false, hls_meta,music)
```
I changed it to this to see if that would do anything. But still, plays to the end, then picks up on new songs.
```
getNew = fallback([requests, music])
radio101 = metadata.map(update=false, hls_meta,getNew)
```


plugins/request.py
```
# ... get the next song info from Firestore
URL = next_song[const.URL]
new_now_playing = now_playing + 1
channel_ref.set({const.NOW_PLAYING : new_now_playing}, merge=True)
print(URL)
```

<details><summary>plugins/dynamic.liq</summary>
<p>

```
def get_request()
  command = process.quote.command("python3", args=["request.py"])
  uri =
    try
      list.hd(process.read.lines(command))
    catch _ do
      null()
    end

  null.map(fun(uri) -> request.create(uri, temporary=true), uri)
end

def replaces request.dynamic(%argsof(request.dynamic), callback)
  prevent = ref(false)

  def disable_prevent()
    prevent := false
  end

  def enable_prevent()
    prevent := true
    delay = max(getter.get(retry_delay) - 0.1, 0.1)
    thread.run(disable_prevent, delay=delay)
  end

  calls = ref(0)
  max_calls = 3

  def reset_calls()
    calls := 0
  end

  def increment_calls()
    calls := calls() + 1
  end

  thread.run(reset_calls, every=retry_delay, fast=true)

  def debounce()
    if prevent() then
      null()
    elsif calls() >= max_calls then
      null()
    else
      result = callback()
      if result == null() then
        enable_prevent()
      end
      increment_calls()
      result
    end
  end

  request.dynamic(debounce, %argsof(request.dynamic))
end
```

</p>
</details>

**Reply by vitoyucepi** [reply]
I don't know what is happening in `plugins/request.py`.

I think you should refactor the `get_request` function like this.
```ruby
def get_request()
  command = process.quote.command("python3", args=["request.py"])
  process = process.run(command)
  
  output = 
    if process.status == "exit" and process.status.code == 0 then
      process.stdout
    else
      null()
    end
  lines = null.map(fun(output) -> r/\r?\n/.split(output), output)
  uri = 
    try
      null.map(list.hd, lines)
    catch _ do
      null()
    end

  null.map(fun(uri) -> request.create(uri, temporary=true), uri)
end
```

**Reply by carrzkiss** [reply]
>I don't know what is happening in plugins/request.py

Is there something wrong with this code? [tristan-nel](https://github.com/tristan-nel) posted this in his original code, which I referenced in this thread.
```
# ... get the next song info from Firestore
URL = next_song[const.URL]
new_now_playing = now_playing + 1
channel_ref.set({const.NOW_PLAYING : new_now_playing}, merge=True)
print(URL)
```

Also, it still does not work.
It only updates when the list is completed.
So, there is something I am missing in all this.

**Reply by vitoyucepi** [reply]
This code snippet tells me *literally nothing*.

**Reply by carrzkiss** [reply]
OK, what are you using for this?
`  command = process.quote.command("python3", args=["request.py"])`
What is your `request.py`

**Reply by vitoyucepi** [reply]
I imagined a script that would output requests one per line, separated by `\n` or `\r\n`. It can output nothing or exit with a non-zero code if an error occurs.

**Reply by carrzkiss** [reply]
OK, I have no idea how to do anything like that.

**Reply by gAlleb** [reply]
Sorry to interfere. But @carrzkiss, are you sure that you need that? How come that you need something without knowing what it is.

If you are not familiar with python, maybe you can implement the required behavior with another language?

`request.dynamic` is used to manage your music via external scripts (which you have to come up with yourself - it could be php API's or bash scripts or python scripts or even js scripts, you name it) or fetch requests from users.  Or like that guy whom you're referring to with this chunk of python script: probably he has made his own solution with Google's Firestore database - but who knows how or what... Do you use Google's Firestore?

To play your music from playlists you can use `playlist` operator (which is basically a function with `request.dynamic` in it already). And what does it do? Dynamically requests tracks from a file with lines in it - your playlist.  When you make new entries in your file or edit it - playlist would be reloaded if it's not remote one. For remote one you've been given solutions like `reload` either with rounds or within time interval.

There is a wonderful book - https://www.liquidsoap.info/book/book.pdf which describes in detail the logic of liquidsoap. While some part of it is more tech-specific, another one is very very easy to comprehend which could help you to build up your station in a way you want.

In regards to the script, that Vito has given you - it is a wonderful LS solution to make request to... `something` and avoid errors or crashing. But you have to understand that `this something` has to be existent and be written for Vito's function to obtain it.

**Reply by carrzkiss** [reply]
Thanks, Stefan.
I am going to get up to speed on what the Python script does and how to use it.
I am thinking of creating another channel and testing out the script [tristan-nel](https://github.com/tristan-nel) provided.
Once I find out what it does and how to properly use it.

I love learning new things and have messed around with Python, but that was about 20 years ago. I have not needed it since then, so I have not learned more about it. But I know that it is used in many products that may come in useful for me in the future, especially with the radio station site, so minus well learn what I need to know to get the job done.

Once I can grasp the other script, I will revisit Vito's script here.

**Comment by carrzkiss** [comment]
[vitoyucepi](https://github.com/vitoyucepi)
I am now getting the chance to get back into this project again, after months away from it.

On this part here.
`command = process.quote.command("python3", args=["request.py"])`

You stated.
> I imagined a script that would output requests one per line, 
> separated by \n or \r\n. It can output nothing or exit with a non-zero code if an error occurs.

This script is this all it does.
```
song1.mp3
song2.mp3
etc...
```
When you move songs around, does the list that resides in request.py update and feed it to Liquidsoap to play in the new order?
```
song2.mp3
song1.mp3
etc...
```
I know how to do the scripting in ASP.NET and Classic ASP.
But I do not think they will work in this situation.

So, I am looking at how to script it in Python with SQL Server.
I can output a song per line.

Will that be all that is needed just to have that script inside of the request.py?
Just a file that is filled with songs that are being ordered by a database or other feed?

Thanks guys.

**Reply by vitoyucepi** [reply]
Hi @carrzkiss,
I don't remember exactly what I did in the code snippet, but it looks like I didn't explain some things.
```ruby
def get_request()
  command = process.quote.command("python3", args=["request.py"])
  process = process.run(command)
  
  output = 
    if process.status == "exit" and process.status.code == 0 then
      process.stdout
    else
      null()
    end
  lines = null.map(fun(output) -> r/\r?\n/.split(output), output)
  uri = 
    try
      null.map(list.hd, lines)
    catch _ do
      null()
    end

  null.map(fun(uri) -> request.create(uri, temporary=true), uri)
end
```
This function creates a single request by extracting the first line from the script output.

**Reply by carrzkiss** [reply]
OK, so your script calls request.py and gets the FIRST line from that file every few seconds; it checks.
So, am I correct on what I need to write in my py script?
A list of songs and your script will then check and grab the song from.

**Reply by vitoyucepi** [reply]
I think it would be better to start a new discussion from scratch.

**Reply by carrzkiss** [reply]
New thread here [4359](https://github.com/savonet/liquidsoap/discussions/4359)
